File: v1/functions/post_roles/methods/get.py
import json
from .contrib.db import db


def get(e={}):
    data = None
    meta = None
    message = None
    status_code = 200

    body = e.get('queryStringParameters') or {}
    secret = body.get('secret')

    if not secret:
        status_code = 403
        message = 'secret is invalid'

        return data, meta, message, status_code

    fields = (
        'post_role.id',
        'post_role.user_id',
        'post_role.grantor_id',
        'post_role.type',
        'post_role.post_key',
        'post_role.secret',
        'post_role.status',
        'post_role.created_at',
        'post_role.updated_at',
    )

    data = (
        db.table('post_role')
        .select(*fields)
        .where('secret', secret)
        .get()
    )

    return data, meta, message, status_code

File: messages/get.js
import _ from 'lodash';
import { select, in as $in, like } from 'sql-bricks';
import table from './lib/table';
import parse, { parseSQLError } from './lib/parse';
import getConnection from './lib/getConnection';

export default async (e, context, callback) => {
  const conn = await getConnection();
  const params = _.get(e, 'queryStringParameters') || {};
  const { message, translation } = table;
  const {
    q,
    key,
    lcid,
  } = params;
  const columns = [
    message.columns.id,
    message.columns.key,
    translation.columns.body,
    translation.columns.lcid,
  ];

  let data = [];
  let sql = select(columns)
    .from(message.name)
    .leftJoin(translation.name, { [translation.columns.messageId]: message.columns.id })
    .orderBy(`${message.columns.createdAt} DESC`);

  if (!_.isNil(q)) {
    const inSQL = select(message.columns.id)
      .from(message.name)
      .join(translation.name, { [translation.columns.messageId]: message.columns.id })
      .where(like(translation.columns.body, `%${q}%`))
      .groupBy(message.columns.id)
      .orderBy(`${message.columns.createdAt} DESC`);

    sql = select(columns)
      .from(message.name)
      .join(translation.name, { [translation.columns.messageId]: message.columns.id })
      .where($in(message.columns.id, inSQL))
      .orderBy(`${message.columns.createdAt} DESC`);
  }

  if (!_.isNil(key)) {
    sql = sql.where(like(message.columns.key, `%${key}%`));
  }

  if (!_.isNil(lcid)) {
    sql = sql.where(translation.columns.lcid, lcid);
  }

  let response;
  const headers = { 'Access-Control-Allow-Origin': '*' };

  try {
    const grouped = _.groupBy(await conn.query(sql.toString()), 'id');
    data = _.map(_.values(grouped), t => parse(t, true));

    response = {
      body: JSON.stringify({ data }),
      headers,
      statusCode: 200,
    };
  } catch (error) {
    response = {
      body: JSON.stringify({ error: parseSQLError(error) }),
      headers,
      statusCode: 400,
    };
  } finally {
    conn.end();
  }

  callback(null, response);
};
